Cloud-init Fails

After adding the iso the first time, the second run fails:



panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x8 pc=0x104837bb0]

goroutine 1 [running]:
github.com/luthermonson/go-proxmox.(*Task).Ping(0x0)
        /Users/mbenjemaa/workspace/go/pkg/mod/github.com/luthermonson/go-proxmox@v0.0.0-alpha3/tasks.go:39 +0x20
github.com/luthermonson/go-proxmox.(*Task).Wait(0x0, 0x3b9aca00, 0x77359400)
        /Users/mbenjemaa/workspace/go/pkg/mod/github.com/luthermonson/go-proxmox@v0.0.0-alpha3/tasks.go:145 +0x2c
github.com/luthermonson/go-proxmox.(*Task).WaitFor(...)
        /Users/mbenjemaa/workspace/go/pkg/mod/github.com/luthermonson/go-proxmox@v0.0.0-alpha3/tasks.go:140
github.com/luthermonson/go-proxmox.(*VirtualMachine).CloudInit(0x140001980e0, {0x104844d81, 0x4}, {0x10485df67, 0x1ed}, {0x10485df67, 0x1ed})
        /Users/mbenjemaa/workspace/go/pkg/mod/github.com/luthermonson/go-proxmox@v0.0.0-alpha3/virtual_machine.go:149 +0x1a8

--- a/proxmox.go
+++ b/proxmox.go
@@ -41,6 +41,12 @@ func IsNotFound(err error) bool {
 	return err == ErrNotFound
 }
 
+var ErrNoop = errors.New("nothing to do")
+
+func IsErrNoop(err error) bool {
+	return err == ErrNoop
+}
+
 func MakeTag(v string) string {
 	return fmt.Sprintf(TagFormat, v)
 }

--- a/virtual_machine.go
+++ b/virtual_machine.go
@@ -18,6 +18,7 @@ const (
 
 	UserDataISOFormat = "user-data-%d.iso"
 	TagCloudInit      = "cloud-init"
+	TagSeperator      = ";"
 
 	volumeIdentifier = "cidata"
 	blockSize        = 2048
@@ -64,7 +65,7 @@ func (v *VirtualMachine) HasTag(value string) bool {
 
 func (v *VirtualMachine) AddTag(value string) (*Task, error) {
 	if v.HasTag(value) {
-		return nil, nil //noop
+		return nil, ErrNoop
 	}
 
 	if v.VirtualMachineConfig.TagsSlice == nil {
@@ -72,7 +73,7 @@ func (v *VirtualMachine) AddTag(value string) (*Task, error) {
 	}
 
 	v.VirtualMachineConfig.TagsSlice = append(v.VirtualMachineConfig.TagsSlice, value)
-	v.VirtualMachineConfig.Tags = strings.Join(v.VirtualMachineConfig.TagsSlice, ";")
+	v.VirtualMachineConfig.Tags = strings.Join(v.VirtualMachineConfig.TagsSlice, TagSeperator)
 
 	return v.Config(VirtualMachineOption{
 		Name:  "tags",
@@ -82,7 +83,7 @@ func (v *VirtualMachine) AddTag(value string) (*Task, error) {
 
 func (v *VirtualMachine) RemoveTag(value string) (*Task, error) {
 	if !v.HasTag(value) {
-		return nil, nil //noop
+		return nil, ErrNoop
 	}
 
 	if v.VirtualMachineConfig.TagsSlice == nil {
@@ -98,7 +99,7 @@ func (v *VirtualMachine) RemoveTag(value string) (*Task, error) {
 		}
 	}
 
-	v.VirtualMachineConfig.Tags = strings.Join(v.VirtualMachineConfig.TagsSlice, ";")
+	v.VirtualMachineConfig.Tags = strings.Join(v.VirtualMachineConfig.TagsSlice, TagSeperator)
 	return v.Config(VirtualMachineOption{
 		Name:  "tags",
 		Value: v.VirtualMachineConfig.Tags,
@@ -106,7 +107,7 @@ func (v *VirtualMachine) RemoveTag(value string) (*Task, error) {
 }
 
 func (v *VirtualMachine) SplitTags() {
-	v.VirtualMachineConfig.TagsSlice = strings.Split(v.VirtualMachineConfig.Tags, ";")
+	v.VirtualMachineConfig.TagsSlice = strings.Split(v.VirtualMachineConfig.Tags, TagSeperator)
 }
 
 // CloudInit takes four yaml docs as a string and make an ISO, upload it to the data store as <vmid>-user-data.iso and will
@@ -145,12 +146,8 @@ func (v *VirtualMachine) CloudInit(device, userdata, metadata, vendordata, netwo
 		return err
 	}
 
-	task, err = v.AddTag(MakeTag(TagCloudInit))
-	if err != nil {
-		return err
-	}
-
-	if err := task.WaitFor(2); err != nil {
+	_, err = v.AddTag(MakeTag(TagCloudInit))
+	if err != nil && !IsErrNoop(err) {
 		return err
 	}
 

